src/lib/error.rs

```diff
@@ -0,0 +1,194 @@
+impl From<std::process::ExitCode> for CargoMakeError {
+    fn from(error: std::process::ExitCode) -> Self {
+        Self::ExitCode(error)
+    }
+}
+
+impl std::process::Termination for CargoMakeError {
+    fn report(self) -> std::process::ExitCode {
+        if let CargoMakeError::ExitCode(exit_code) = self {
+            return exit_code;
+        }
+        let status_code = self.discriminant();
+        if status_code > u8::MAX as u16 {
+            eprintln!("exit code {}", status_code);
+            std::process::ExitCode::FAILURE
+        } else {
+            std::process::ExitCode::from(status_code as u8)
+        }
+    }
+}
+
+pub enum SuccessOrCargoMakeError<T> {
+    Ok(T),
+    Err(CargoMakeError),
+}
+
+impl<T> From<Result<T, CargoMakeError>> for SuccessOrCargoMakeError<T> {
+    fn from(value: Result<T, CargoMakeError>) -> Self {
+        match value {
+            Ok(val) => SuccessOrCargoMakeError::Ok(val),
+            Err(error) => SuccessOrCargoMakeError::Err(error),
+        }
+    }
+}
+
+// Can't use `Result` because
+// [E0117] Only traits defined in the current crate can be implemented for arbitrary types
+impl<T> std::process::Termination for SuccessOrCargoMakeError<T> {
+    fn report(self) -> std::process::ExitCode {
+        match self {
+            SuccessOrCargoMakeError::Ok(_) => std::process::ExitCode::SUCCESS,
+            SuccessOrCargoMakeError::Err(error) => {
+                if let CargoMakeError::StdIoError { ref error } = error {
+                    if let Some(e) = error.raw_os_error() {
+                        eprintln!("{}", e.to_string());
+                        return if e > u8::MAX as i32 {
+                            eprintln!("exit code {}", e);
+                            std::process::ExitCode::FAILURE
+                        } else {
+                            std::process::ExitCode::from(e as u8)
+                        };
+                    }
+                }
+                eprintln!("{}", error.to_string());
+                error.report()
+            }
+        }
+    }
+}
+
+/*
+// TODO: Get `Into<CargoMakeError>` syntax working
+impl std::ops::FromResidual<Result<std::convert::Infallible, CargoMakeError>>
+for SuccessOrCargoMakeError<std::process::ExitCode>
+{
+    fn from_residual(residual: Result<std::convert::Infallible, CargoMakeError>) -> Self {
+        if let Err(e) = residual {
+            SuccessOrCargoMakeError::Err(e)
+        } else {
+            SuccessOrCargoMakeError::Ok(std::process::ExitCode::SUCCESS)
+        }
+    }
+}
+
+impl std::ops::FromResidual<Result<std::convert::Infallible, std::io::Error>>
+for SuccessOrCargoMakeError<std::process::ExitCode>
+{
+    fn from_residual(residual: Result<std::convert::Infallible, std::io::Error>) -> Self {
+        if let Err(e) = residual {
+            SuccessOrCargoMakeError::Err(e.into())
+        } else {
+            SuccessOrCargoMakeError::Ok(std::process::ExitCode::SUCCESS)
+        }
+    }
+}
+*/
```

Comment:
what is this todo about?

Reply:
If you're ok with enabling `#![feature(try_trait_v2)]` then I will put that in, uncomment this, and use `?` in your `main` entrypoints.

Reply:
not familiar with it but i want to be backward compatible as much as possible including in rust versions. so lets drop it for now or if we can wrap it with some config that checks if feature is enabled?

Reply:
Ok it's nightly only still it seems https://github.com/metalbear-co/mirrord/issues/1568 https://github.com/rust-lang/rust/issues/84277 so will remove the commented out lines. This PR is ready to merge.